connector/src/main/scala/com/microsoft/kusto/spark/datasink/KustoWriter.scala

```diff
@@ -192,66 +195,75 @@ object KustoWriter {
                                                     batchIdIfExists: String,
                                                     kustoAdminClient: Client,
                                                     tmpTableName: String,
+                                                    acc: CollectionAccumulator[PartitionResult],
+                                                    timeout: FiniteDuration,
                                                     isAsync: Boolean = false): Unit = {
     import coordinates._
-    try {
-      // Move data to real table
-      // Protect tmp table from merge/rebuild and move data to the table requested by customer. This operation is atomic.
-      kustoAdminClient.execute(database, generateTableAlterMergePolicyCommand(tmpTableName, allowMerge = false, allowRebuild = false))
-      kustoAdminClient.execute(database, generateTableMoveExtentsCommand(tmpTableName, table.get))
-      KDSU.logInfo(myName, s"write to Kusto table '${table.get}' finished successfully $batchIdIfExists")
-    }
-    catch {
-      case exception: Exception =>
-        KDSU.reportExceptionAndThrow(myName, exception, "finalizing write operation", cluster, database, table.get, isLogDontThrow = isAsync)
-    }
-    finally {
-      tryCleanupIngestionByproducts(database, kustoAdminClient, tmpTableName)
-    }
-  }
 
-  private def ingestToTemporaryTableByWorkers(
-                                               batchId: String,
-                                               timeout: FiniteDuration,
-                                               storageUri: String,
-                                               rows: Iterator[InternalRow])
-                                             (implicit parameters: KustoWriteResource)
-  : Unit = {
-
-    import parameters._
-    val partitionId = TaskContext.getPartitionId
-    KDSU.logInfo(myName, s"Ingesting partition '$partitionId'")
-    val targetTable = coordinates.table.get
-    val ingestClient = KustoClientCache.getClient(coordinates.cluster, authentication).ingestClient
-
-    // We force blocking here, since the driver can only complete the ingestion process
-    // once all partitions are ingested into the temporary table
-    Await.result(
-      Future(ingestRowsIntoKusto(rows, storageUri, ingestClient)).map { ingestionResults =>
-        // Proceed only on success. Will throw on failure for the driver to handle
-        ingestionResults.foreach {
-          ingestionResult =>
+    val partitionsResults = acc.value
+    val mergeTask = Future {
+      try {
+        partitionsResults.asScala.foreach {
+          partitionResult =>
             KDSU.runSequentially[IngestionStatus](
-              func = () => ingestionResult.getIngestionStatusCollection.get(0),
+              func = () => partitionResult.ingestionResult.getIngestionStatusCollection.get(0),
```

Comment:
This logic should be at KustoClient.
Each table call is independent of the others, therefore I don't see any value of merging them all together.
Let's expose at KustoCLient a method called: waitForResultToComplete, which will return Future, and then we can wait on all of those.